drv/sidecar-mainboard-controller/src/tofino2.rs

```diff
@@ -131,117 +132,38 @@ pub struct TofinoSeqAbort {
     pub error: TofinoSeqError,
 }
 
-#[derive(Copy, Clone, Debug, Eq, PartialEq)]
-pub enum PowerRails {
-    Vdd18,
-    VddCore,
-    VddPcie,
-    Vddt,
-    Vdda15,
-    Vdda18,
-}
-
-#[derive(
-    Copy, Clone, Debug, Default, Eq, PartialEq, FromPrimitive, AsBytes,
-)]
-#[repr(u8)]
-pub enum PowerRailState {
-    #[default]
-    Disabled = 0,
-    RampingUp = 1,
-    GoodTimeout = 2,
-    Aborted = 3,
-    Enabled = 4,
-}
-
-impl TryFrom<u8> for PowerRailState {
-    type Error = FpgaError;
-
-    fn try_from(v: u8) -> Result<Self, Self::Error> {
-        Self::from_u8(v).ok_or(FpgaError::InvalidValue)
-    }
-}
-
-#[derive(Copy, Clone, Debug, Default, Eq, PartialEq)]
+#[derive(Copy, Clone, Debug, Eq, FromPrimitive, PartialEq)]
 #[repr(C)]
-pub struct PowerRailPinState {
-    pub enable: bool,
-    pub good: bool,
-    pub fault: bool,
-    pub vrhot: bool,
-}
-
-impl From<u8> for PowerRailPinState {
-    fn from(v: u8) -> Self {
-        use Reg::TOFINO_POWER_VDD18_STATE::*;
-
-        Self {
-            enable: v & ENABLE != 0,
-            good: v & GOOD != 0,
-            fault: v & FAULT != 0,
-            vrhot: v & VRHOT != 0,
-        }
-    }
+pub enum TofinoPowerRailId {
+    Vdd18 = 0,
+    VddCore = 1,
+    VddPcie = 2,
+    Vddt = 3,
+    Vdda15 = 4,
+    Vdda18 = 5,
 }
 
 #[derive(Copy, Clone, Debug, Eq, PartialEq)]
 #[repr(C)]
-pub struct PowerRail {
-    pub id: PowerRails,
-    pub state: PowerRailState,
-    pub pin_state: PowerRailPinState,
+pub struct TofinoPowerRail {
+    pub id: TofinoPowerRailId,
+    pub status: PowerRailStatus,
+    pub pins: PowerRailPinState,
 }
 
-impl TryFrom<(PowerRails, u8)> for PowerRail {
+impl TryFrom<(usize, PowerRailState)> for TofinoPowerRail {
     type Error = FpgaError;
 
-    fn try_from(data: (PowerRails, u8)) -> Result<Self, Self::Error> {
-        let (id, v) = data;
-        Ok(PowerRail {
-            id,
-            state: PowerRailState::try_from(v >> 4)?,
-            pin_state: PowerRailPinState::from(v & 0xf),
+    fn try_from(data: (usize, PowerRailState)) -> Result<Self, Self::Error> {
```

Comment:
I'm not convinced that `impl TryFrom` is better than
```rust
impl TofinoPowerRail {
    fn new(id: usize, state: PowerRailState) -> Result<Self, FpgaError> {
        // ...
    }
}
```
since constructing / indexing the tuple is awkward.

Reply:
I am hesitant here. I understand what you're saying, but conceptually I believe that a `new(..)` function should not error. This may be a left over from my C++ days where throwing exceptions from a constructor was problematic, but I still like to separate memory allocation/initialization from functions which may cause errors. As such I feel like the `try_from(..)`, despite its awkward definition better conveys that we're doing something which has a change of failing. Is there a potential way/trait/pattern that may satisfy both objectives?

Reply:
Having `new(...)` be fallible isn't _totally_ unheard of, e.g. [`NonZeroU8::new`](https://doc.rust-lang.org/std/num/struct.NonZeroU8.html#method.new) returns an `Option<Self>`.

It does appear to be uncommon, though.  What about making an infallible `new`
```rust
impl TofinoPowerRail {
    fn new(id: TofinoPowerRailId, status: PowerRailStatus, pins: PowerRailPinState) -> Self {
        // ...
    }
}
```
then doing the type conversions in the loop at [`tofino2.rs:292`](https://github.com/oxidecomputer/hubris/pull/1449/files#diff-580ceb53b19ee484f37faf79894c68765ef0a5ed11d39fa7b9d50641f2c3ce41R293)?  That appears to be the only place `try_from` is used.

Reply:
Alright, fair point. This only used in one place so we may as well pull the loop into that function. There's no need really to implement this in a separate `TryFrom`.